### ISSUE TITLE
Docker container Bind Mount strange behavior

### ISSUE THREAD

**Issue by fabiom-zz** [MicrosoftDocs/vsonline#433]
I would like to allow docker containers inside a VS online environment to access the workspace files via Docker Bind Mounts. I think this is a valid use case in development scenarios.

However, the behavior in VS Online environments is strange and unexpected.

Steps to reproduce:

* Create an empty VS Online environment
* The workspace files will be located under `/home/vsonline/workspace`
* When I try to create a Docker container with a Bind Mount mapped to `/home/vsonline/workspace` using:

`sudo docker run -d -it --name testmount --mount type=bind,source="/home/vsonline",target=/app nginx:latest`

I get the error:

`docker: Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /home/vsonline/workspace.`

which is very odd since the directory does exist.

* The same error also happens to `/home/vsonline`. However, mapping to `/home` does NOT generate an error message:

`sudo docker run -d -it --name testmount --mount type=bind,source="/home",target=/app nginx:latest`

But what the Docker container actually sees is totally strange. I log into the container via:

`sudo docker exec -it testmount /bin/bash`

Once in the container, the content of `/app` that maps to `/home` is totally different.

```
root@1992bd1dfa10:/# ls /app
cloudenv
```

The directory `/home` in the host doesn't have the `cloudenv` directory. It looks very different.

Actually I suspect there is some behind the scenes stuff going on. When I use -v (which creates missing directories automatically) instead of --mount (which does not) I do NOT get an error.

And more interesting: I noticed *files are created and persisted among different containers*. I just have no idea where they are :-)

Bottom line: how can I map the workspace directory inside a Docker container inside the environment?

**Comment by edgonmsft** [comment]
Hi fabiom-zz! 

Could you also let us know more about your use case?

For now let me point out the following:

The way containers are implemented right now is as Docker outside Docker. Your environment is a container but you are taking to the docker instance that manages it. You can verify this by doing docker ps -a.

That is the reason the paths don't match since you would have to mount the folder that is mounted in your current container.

The current best solution for this is docker-compose.yml file to set this up. We just shipped initial support for it.

You can follow this example: 
https://github.com/microsoft/vscode-dev-containers/tree/master/containers/python-3-postgres/.devcontainer

The main change would be to reference the workspace volume in both of your containers.

Let us know what you think!
Thanks!

**Comment by fabiom-zz** [comment]
One of my use cases is a Jenkins in Docker . We build scripts that act on the local workspace files. We could install Jenkins directly in the environment to work around. 

It seems that spinning Docker containers with specific tech/frameworks (like the Python/Postgres example you shared) that read directly from the workspace is another good example.

Do we need to use Docker Compose to get this working? I could not get the mapping working using plain `docker run` commands.

`sudo docker run -d -it --name testmount --mount type=bind,source="..",target=/app nginx:latest`

Does not work because the mount path must be absolute.

`sudo docker run -d -it --name testmount --mount type=bind,source="/home/vsonline/workspace",target=/app nginx:latest`

Does not work because Docker reports the source path does not exist.

What would be the `docker run` command that would work?

**Comment by edgonmsft** [comment]
Hi fabiom-zz!

Sorry for the delay! I thought I had already answered your question.

If you run docker inspect on the container where your environment is currently running you can see the path we are using to mount the workspace and reuse it in your new containers.

We will probably change this path in the future to use volumes which would be easier to reuse in the future, but for now that should unblock you.

**Comment by edgonmsft** [comment]
Closing. If you have any more issues feel free to reopen or open a new issue.

Thanks!

**Comment by orph** [comment]
@edgonmsft I'm having the following error running docker-in-docker, any ideas?

```
~/workspace# docker run -ti  ubuntu
docker: Error response from daemon: error creating aufs mount to /var/lib/docker/aufs/mnt/9e3ff4c90eef43bc006272944e4dab6483934ac5bcdaf4d5d6278cc5852336af-init: mount target=/var/lib/docker/aufs/mnt/9e3ff4c90eef43bc006272944e4dab6483934ac5bcdaf4d5d6278cc5852336af-init data=br:/var/lib/docker/aufs/diff/9e3ff4c90eef43bc006272944e4dab6483934ac5bcdaf4d5d6278cc5852336af-init=rw:/var/lib/docker/aufs/diff/8a7a89868790ed45ceac87ca0fe0153ff63728abcb3229a371867bc470f820c5=ro+wh:/var/lib/docker/aufs/diff/e0a0dfe8cfcda6535d12e0f3f4687523f5c1c0c0ab86389483d3ad2cd1507fbe=ro+wh:/var/lib/docker/aufs/diff/4ffef8a0ac57f11e3423e7d91421fce2c979dfcfb13f079e566cdf8ea4d9b925=ro+wh:/var/lib/docker/aufs/diff/1b70b1fba877adb33885d8fb896859548370be708193c51359caf3fe59c569ce=ro+wh,dio,xino=/dev/shm/aufs.xino: invalid argument.
See 'docker run --help'.
```

**Comment by edgonmsft** [comment]
Hi Alex!

Can you give me more information on how you created the environment you are trying it from? `devcontainer.json` and the image you are using?

**Comment by orph** [comment]
Pretty much just a vanilla Debian stretch with docker-ce installed. dockerd
is run without options in the background by supervisord.

**Comment by orph** [comment]
@edgonmsft Pretty much just a vanilla Debian stretch with docker-ce installed. dockerd is run without options in the background by supervisord.

**Comment by orph** [comment]
@edgonmsft for more context, the devcontainer.json uses 
```
	"dockerFile": "Dockerfile",
	"context": "./../",
	"runArgs": ["--privileged", "--security-opt", "seccomp=unconfined"],
	"overrideCommand": false,
```

And the dockerfile installs docker this way:
```
FROM debian:stretch
RUN curl --location --verbose "https://download.docker.com/linux/debian/gpg" | apt-key add - && \
	echo "deb [arch=amd64] https://download.docker.com/linux/debian stretch stable" > /etc/apt/sources.list.d/docker.list && \
	apt-get update && \
	apt-get install --yes docker-ce
```

**Comment by orph** [comment]
Fixed my issue by adding `VOLUME /var/lib/docker` to my Dockerfile. Nested dockers with volume mounts in the parent docker working as expected now!

**Comment by edgonmsft** [comment]
Sorry for the delay Alex!

That's good to know!

**Comment by orph** [comment]
I wonder if this volume mount should be added by default? I haven’t seen it
documented anywhere, and it’s not obvious. I just happened to see it being
used In our CI dockers.

**Comment by edgonmsft** [comment]
Good question. @Chuxel any thoughts?

**Comment by Chuxel** [comment]
@edgonmsft @orph Sorry which volume mount?  The docker executable one?  That's actually not recommended in the same article that describes general problems with docker-in-docker.   I do not think we should be doing it but rather recommending the installing the CLI in the image which is easy enough to do.

**Comment by orph** [comment]
@Chuxel huh! This is a parent docker with docker-ce installed and dockerd running. You still can’t run docker-in-docker with volume mounts without adding `VOLUME /bar/lib/docker` in the parent’s Dockerfile - I assume because aufs doesn’t support layering.

**Comment by Chuxel** [comment]
@orph Yeah, the specific problem is bind mounting the local VM's `/var/lib/docker` - it can cause data corruption. Creating a named volume largely un-does that... which is why I was asking @edgonmsft if we bind mount it by default.

True docker-in-docker is really problematic given the number of pitfalls. I think the thing we're primarily missing to get the outer docker working with bind mounts is to enable `${localWorkspaceFolder}` so an env var can be set.  We'd want to document it as a way to deal with it.  e.g. you can do thinks like ${LOCAL_WORKSPACE_FOLDER:-.} which would use the current folder if the env var is not set.

**Comment by edgonmsft** [comment]
@Chuxel We are not bind mounting /var/lib/docker to the container or anything. Only the socket and executable by default and only in default image environments. In other cases we take the socket mount if its included and the user would have to install the cli in their container/image.

Can you give me more details on how we would like to make `${localWorkspaceFolder}` work? The idea would be that its set on the container so that when you create containers from within it the folder on the vm that has the workspace can be bind mounted to other containers?

**Comment by Chuxel** [comment]
@edgonmsft Yep! We doc this for Remote - Containers. https://code.visualstudio.com/docs/remote/containers-advanced#_mounting-host-volumes-with-docker-from-inside-a-container
